### ISSUE TITLE
command(s?) in documentation don't work. 

### ISSUE THREAD

**Issue by mikedlr** [kubernetes-sigs/cloud-provider-kind#144]
In the main documentation there is a command

    $ kubectl label node kind-control-plane node.kubernetes.io/exclude-from-external-load-balancers-
    node/kind-control-plane unlabeled

This command does not work. I believe that it should be something like

    kubectl label node --overwrite kind-2-control-plane node.kubernetes.io/exclude-from-external-load-balancers=true

instead - probably better if the prompt isn't included in the copied text.  It might be good if the documents were automatically tested e.g. with https://github.com/JNRowe/shell-doctest

**Comment by BenTheElder** [comment]
> It might be good if the documents were automatically tested e.g. with https://github.com/JNRowe/shell-doctest

that might get a bit expensive since you'd need the requisites setup.

they're also probably not changing often.

I agree that we should have copyable commands.

**Comment by zrisher** [comment]
I also personally felt the documentation around this command was unclear:

> Control-plane nodes need to remove the special label node.kubernetes.io/exclude-from-external-load-balancers to be able to access the workloads running on those nodes using a LoadBalancer Service.

Does "these nodes" mean just the control-plane node? Or the nodes it controls? I.e. do I need to run this command if I don't plan on putting a load balancer in front of my control plane?

**Comment by aojea** [comment]
@zrisher how familiar are you with kubernetes? just trying to understand if you are looking at this from a totally new person or with some background

**Comment by zrisher** [comment]
First K8S project, a lot of hours in but also many new technologies to research besides just K8S. Definitely the first time I've tried to get a load balancer running on a local cluster.

**Comment by aojea** [comment]
yeah, I see, this project README assumes some previous knowledge, now I understand why you feel this friction

**Comment by zrisher** [comment]
If this configuration step is needed in all cases, I think

> WE also need to remove the label `node.kubernetes.io/exclude-from-external-load-balancers` from the control plane node:

is more clear for both the experienced and the iniates.

**Comment by BenTheElder** [comment]
FWIW I personally think it's generally best practice that _examples_ have `$` and output and "you should run this" only have exactly the command to copy/paste and I would support a PR to correct that.

---

> Does "these nodes" mean just the control-plane node? Or the nodes it controls? I.e. do I need to run this command if I don't plan on putting a load balancer in front of my control plane? Or is this something I need to do to use this tool at all?

There can be more than one control-plane node, which is why it says "these nodes". The nodes with this label.
That label controls excluding nodes from load balancing.

Regarding control plane vs worker nodes ...  https://kind.sigs.k8s.io/docs/user/configuration/#nodes

> If this configuration step is needed in all cases, I think


It is not required to use the tool. You could only serve traffic to other nodes. It is dependent on your use case.

--- 

The official docs for this label are at https://kubernetes.io/docs/reference/labels-annotations-taints/ which we should reference. Specifically https://kubernetes.io/docs/reference/labels-annotations-taints/#node-kubernetes-io-exclude-from-external-load-balancers

The more specific reference is: https://kubernetes.io/docs/setup/production-environment/tools/kubeadm/create-cluster-kubeadm/#control-plane-node-isolation

**Comment by zrisher** [comment]
Thanks @BenTheElder, this clears things up for me a bit. I had read all of this except for that last link - the source of my confusion was that I assumed people would not be scheduling workloads onto their control-plane node. So in the case I envisioned, if the docs say we need to remove `node.kubernetes.io/exclude-from-external-load-balancers` from the control plane node, it must somehow affect nodes managed by that control plane.

In fact, as I understand now, you *only* need to remove this label if you are trying to loadbalance a workload served by a pod that is scheduled onto a control plane node (which is a case I didn't account for).

Given that understanding, I would perhaps suggest:

> If you are running workloads on control plane nodes (as is the [default kind configuration](https://kind.sigs.k8s.io/docs/user/configuration/#nodes)), you will need to remove the special label [`node.kubernetes.io/exclude-from-external-load-balancers`](https://kubernetes.io/docs/reference/labels-annotations-taints/#node-kubernetes-io-exclude-from-external-load-balancers) to access those workloads using a LoadBalancer.

**Comment by BenTheElder** [comment]
> In fact, as I understand now, you only need to remove this label if you are trying to loadbalance a workload served by a pod that is scheduled onto a control plane node (which is a case I didn't account for).

Yes, that's exactly it. We find users doing all sorts of things with the tools 😅 

---- 

For single-node-clusters we currently automatically remove the scheduling taint that limits the node to only control plane workloads, we should consider automatically removing this label as well. I'll defer to @aojea on that one though, we can track a feature in kubernetes-sigs/kind if we think that makes sense.


The note would still be relevant in some form for people with other cluster shapes though.

Getting robust cross-platform loadbalancing w/ kind is still a bit new (for a long time we pointed to metallb which worked pretty well on Linux at least), and the first use case is the same as `kind`: [developing Kubernetes itself](https://kind.sigs.k8s.io/docs/contributing/project-scope/#p0-support-testing-kubernetes). We clearly have more to do to flesh things out to be friendly for general purpose (which is definitely supported, but needs more work like this issue). 

Thanks for understanding and working with us :-)

**Comment by aojea** [comment]
> Given that understanding, I would perhaps suggest:
> 
> > If you are running workloads on control plane nodes (as is the [default kind configuration](https://kind.sigs.k8s.io/docs/user/configuration/#nodes)), you will need to remove the special label [`node.kubernetes.io/exclude-from-external-load-balancers`](https://kubernetes.io/docs/reference/labels-annotations-taints/#node-kubernetes-io-exclude-from-external-load-balancers) to access those workloads using a LoadBalancer.

SGTM please go ahead and file a PR
